docs/coding-guidelines/vectorization-guidelines.md

```diff
@@ -0,0 +1,1091 @@
+
+```cs
+public static Vector128<T> BitwiseAnd<T>(Vector128<T> left, Vector128<T> right) where T : struct => left & right;
+public static Vector128<T> BitwiseOr<T>(Vector128<T> left, Vector128<T> right) where T : struct => left | right;
+public static Vector128<T> Xor<T>(Vector128<T> left, Vector128<T> right) => left ^ right;
+public static Vector128<T> Negate<T>(Vector128<T> vector) => ~vector;
+```
+
+`AndNot` computes the bitwise-and of a given vector and the ones' complement of another vector.
+
+```cs
+public static Vector128<T> AndNot<T>(Vector128<T> left, Vector128<T> right) => left & ~right;
+```
+
+`ShiftLeft` shifts each element of a vector left by the specified number of bits.
+`ShiftRightArithmetic` performs a **signed** shift right and `ShiftRightLogical` performs an **unsigned** shift:
+
+```cs
+public static Vector128<sbyte> ShiftLeft(Vector128<sbyte> vector, int shiftCount);
+public static Vector128<sbyte> ShiftRightArithmetic(Vector128<sbyte> vector, int shiftCount);
+public static Vector128<byte> ShiftRightLogical(Vector128<byte> vector, int shiftCount);
+```
+
+### Equality
+
+`EqualsAll` compares two vectors to determine if all elements are equal. `EqualsAny` compares two vectors to determine if any elements are equal.
+
+```cs
+public static bool EqualsAll<T>(Vector128<T> left, Vector128<T> right) where T : struct => left == right;
+public static bool EqualsAny<T>(Vector128<T> left, Vector128<T> right) where T : struct
+```
+
+`Equals` compares two vectors to determine if they are equal on a per-element basis. It returns a vector whose elements are all-bits-set or zero, depending on whether the corresponding elements in the `left` and `right` arguments were equal.
+
+```cs
+public static Vector128<T> Equals<T>(Vector128<T> left, Vector128<T> right) where T : struct
+```
+
+How do we calculate the index of the first match? Let's take a closer look at the result of following equality check:
+
+```cs
+Vector128<int> left = Vector128.Create(1, 2, 3, 4);
+Vector128<int> right = Vector128.Create(0, 0, 3, 0);
+Vector128<int> equals = Vector128.Equals(left, right);
+Console.WriteLine(equals);
+```
+
+```log
+<0, 0, -1, 0>
+```
+
+`-1` is just `0xFFFFFFFF` (all-bits-set). We could use `GetElement` to get the first non-zero element.
+
+```cs
+public static T GetElement<T>(this Vector128<T> vector, int index) where T : struct
+```
+
+But it would not be an optimal solution. We should instead extract the most significant bits:
+
+```cs
+uint mostSignificantBits = equals.ExtractMostSignificantBits();
+Console.WriteLine(Convert.ToString(mostSignificantBits, 2).PadLeft(32, '0'));
+```
+
+```log
+00000000000000000000000000000100
+```
+
+and use [BitOperations.TrailingZeroCount](https://learn.microsoft.com/dotnet/api/system.numerics.bitoperations.trailingzerocount) to get the trailing zero count.
```

Comment:
In .NET 7+, you can also simply use `int.TrailingZeroCount(x)`. It is a bit more accessible and there are corresponding functions on `uint`, `long`, `ulong`, and the other primitive integer types.

The same goes for `LeadingZeroCount` and a few other APIs introduced as part of generic math.